gemrb/core/Streams/FileStream.h

```diff
@@ -98,6 +99,93 @@ struct File {
 		return !fseek(file, offset, SEEK_END);
 	}
 };
+#else
+struct File {
+private:
+	HANDLE file = INVALID_HANDLE_VALUE;
+public:
+	explicit File(FILE* f) : file(f) {}
+	File() noexcept = default;
+	File(const File&) = delete;
+	File(File&& f) noexcept {
+		file = f.file;
+		f.file = INVALID_HANDLE_VALUE;
+	}
+	~File() {
+		if (file != INVALID_HANDLE_VALUE) {
+			CloseHandle(file);
+		}
+	}
+
+	File& operator=(const File&) = delete;
+	File& operator=(File&& f) noexcept {
+		if (&f != this) {
+			std::swap(file, f.file);
+		}
+		return *this;
+	}
+
+	strpos_t Length() {
+		LARGE_INTEGER fileSize;
+		GetFileSizeEx(file, &fileSize);
+
+		return fileSize.QuadPart;
+	}
+	bool OpenRO(const path_t& name) {
+		file = _OpenFile(name, GENERIC_READ, OPEN_EXISTING);
+		return file != INVALID_HANDLE_VALUE;
+	}
+	bool OpenRW(const path_t& name) {
+		file = _OpenFile(name, GENERIC_READ | GENERIC_WRITE, OPEN_EXISTING);
+		return file != INVALID_HANDLE_VALUE;
+	}
+	bool OpenNew(const path_t& name) {
+		file = _OpenFile(name, GENERIC_WRITE, CREATE_ALWAYS);
+		return file != INVALID_HANDLE_VALUE;
+	}
+	strret_t Read(void* ptr, size_t length) {
+		DWORD bytesRead = 0;
+		ReadFile(file, ptr, length, &bytesRead, nullptr);
+		return bytesRead;
+	}
+	strret_t Write(const void* ptr, strpos_t length) {
+		DWORD bytesWritten = 0;
+		WriteFile(file, ptr, length, &bytesWritten, nullptr);
+		return bytesWritten;
+	}
+	bool SeekStart(stroff_t offset) {
+		return _SetFilePointer(offset, FILE_BEGIN);
+	}
+	bool SeekCurrent(stroff_t offset) {
+		return _SetFilePointer(offset, FILE_CURRENT);
+	}
+	bool SeekEnd(stroff_t offset) {
+		return _SetFilePointer(offset, FILE_END);
+	}
+
+private:
+	HANDLE _OpenFile(const path_t& name, DWORD access, DWORD disposition) {
+		auto fileName = StringFromUtf8(name.c_str());
+
+		return CreateFile(
+			reinterpret_cast<const wchar_t*>(fileName.c_str()),
+			access,
+			FILE_SHARE_READ | FILE_SHARE_WRITE, // some files are opened r and r+w in parallel
+			nullptr,
+			disposition,
+			FILE_ATTRIBUTE_NORMAL,
+			nullptr
+		);
+	}
+
+	bool _SetFilePointer(stroff_t offset, DWORD method) {
+		LARGE_INTEGER liOffset;
+		liOffset.QuadPart = offset;
+
+		return SetFilePointerEx(file, liOffset, nullptr, method) != 0;
+	}
+};
+#endif
```

Comment:
Was there any good reason to have this `File` thing completely in the header? For now I ifdef'ed that into the Windows sction but that also increases the amount of compiler noise and time.

Reply:
I doubt it and yeah, let's change it. I suggest we split out a File.h/File.cpp and add the Windows versions under platforms/. If there are many places that need it, maybe also a shell header, so the ifdefing doesn't have to be repeated too much.